juju/model.py

```diff
@@ -0,0 +1,167 @@
+import json
+from collections.abc import Mapping
+from functools import lru_cache
+from subprocess import run, PIPE
+
+
+class Model:
+    def __init__(self, relation_names, backend):
+        self.relations = RelationMap(relation_names, backend)
+        self._backend = backend
+
+    @property
+    def app(self):
+        return self._backend.local_app
+
+    @property
+    def unit(self):
+        return self._backend.local_unit
+
+    def relation(self, relation_name):
+        """Return the first Relation object for the named relation, or None.
+
+        This is a convenience method for the case where only a single related
+        app is supported. If there are more than one remote apps on the relation,
+        it will raise a TooManyRelatedApps model error. If no there are no remote
+        apps on the relation, it will return None. Otherwise, it is equivalent
+        to self.relations[relation_name][0].
+        """
+        num_related = len(self.relations[relation_name])
+        if num_related > 1:
+            raise TooManyRelatedApps(relation_name, num_related, 1)
+        elif num_related == 0:
+            return None
+        else:
+            return self.relations[relation_name][0]
+
+
+class Application:
+    def __init__(self, name):
+        self.name = name
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+
+class Unit:
+    def __init__(self, name, app):
+        self.name = name
+        self.app = app
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+
+class RelationMap(Mapping):
+    """Map of relation names to lists of Relation instances."""
+    def __init__(self, relation_names, backend):
+        self._relation_names = relation_names
+        self._backend = backend
+
+    def __iter__(self):
+        return iter(self._relation_names)
+
+    def __len__(self):
+        return len(self._relation_names)
+
+    def __getitem__(self, relation_name):
+        return self._backend.get_relations(relation_name)
+
+
+class Relation:
+    def __init__(self, relation_name, relation_id, backend):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend
+        self.data = RelationData(self.relation_name, relation_id, self._backend)
+
+    @property
+    def apps(self):
+        return [unit.app for unit in self.units]
+
+    @property
+    def units(self):
+        return list(self.data.keys())
+
+
+class RelationData(Mapping):
+    def __init__(self, relation_name, relation_id, backend):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend
+        self._keys = self._backend.get_relation_members(self.relation_id)
```

Comment:
Haven't gone through all of it yet, but there might be something to tune here: it sounds awkward to account units and apps as both being "members" of a relation, if that's what this function does. It would be nice to somehow make the API more obvious in this sense. The data bag is heterogeneous and can hold data for units and for apps, but we shouldn't have that kind of mix go unnoticed elsewhere.

Reply:
Part of the refactor I did to remove a bunch of duplication of logic that I noticed when adding the classes for config and leadership cleans this up. If this needs to be cleaned up prior to this PR landing, I could pull in just the refactor bit without any of the new additions to see if it addresses your concerns.